handlers_global.go

```diff
@@ -22,95 +23,171 @@ func (ch contextHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-			log.WithError(err).Debug("Could not extract span from request")
-			span = tracer.StartSpan("/import", trace.NameTag("veneur.opentracing.import")).(*trace.Span)
-		} else {
-			log.WithField("trace", span.Trace).Debug("Extracted span from request")
-		}
-		defer span.Finish()
-
-		innerLogger := log.WithField("client", r.RemoteAddr)
-
-		switch encLogger := innerLogger.WithField("encoding", encoding); encoding {
-		case "":
-			body = r.Body
-			encoding = "identity"
-		case "deflate":
-			body, err = zlib.NewReader(r.Body)
-			if err != nil {
-				http.Error(w, err.Error(), http.StatusBadRequest)
-				span.Error(err)
-				encLogger.WithError(err).Error("Could not read compressed request body")
-				s.statsd.Count("import.request_error_total", 1, []string{"cause:deflate"}, 1.0)
-				return
-			}
-			defer body.Close()
-		default:
-			http.Error(w, encoding, http.StatusUnsupportedMediaType)
-			span.Error(errors.New("Could not determine content-encoding of request"))
-			encLogger.Error("Could not determine content-encoding of request")
-			s.statsd.Count("import.request_error_total", 1, []string{"cause:unknown_content_encoding"}, 1.0)
 			return
 		}
+		// the server usually waits for this to return before finalizing the
+		// response, so this part must be done asynchronously
+		go s.ImportMetrics(span.Attach(ctx), jsonMetrics)
+	})
+}
+
+func handleTraceRequest(ctx context.Context, stats *statsd.Client, w http.ResponseWriter, r *http.Request) (*trace.Span, []DatadogTraceSpan, error) {
+	var (
+		traces []DatadogTraceSpan
+		err    error
+		span   *trace.Span
+	)
+
+	span, err = tracer.ExtractRequestChild("/spans", r, "veneur.opentracing.spans")
+	if err != nil {
+		log.WithError(err).Info("Could not extract span from request")
+		span = tracer.StartSpan("/spans", trace.NameTag("veneur.opentracing.spans")).(*trace.Span)
+	} else {
+		log.WithField("trace", span.Trace).Info("Extracted span from request")
+	}
+	defer span.Finish()
+
+	innerLogger := log.WithField("client", r.RemoteAddr)
+
+	if err = json.NewDecoder(r.Body).Decode(&traces); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		span.Error(err)
+		innerLogger.WithError(err).Error("Could not decode /spans request")
+		stats.Count("import.request_error_total", 1, []string{"cause:json"}, 1.0)
+		return nil, nil, err
+	}
+
+	if len(traces) == 0 {
+		const msg = "Received empty /spans request"
+		http.Error(w, msg, http.StatusBadRequest)
+		span.Error(errors.New(msg))
+		innerLogger.WithError(err).Error(msg)
+		return nil, nil, err
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+	stats.TimeInMilliseconds("spans.response_duration_ns",
+		float64(time.Since(span.Start).Nanoseconds()),
+		[]string{"part:request"},
+		1.0)
 
-		if err = json.NewDecoder(body).Decode(&jsonMetrics); err != nil {
+	return span, traces, nil
+}
+
+func handleRequest(ctx context.Context, stats *statsd.Client, w http.ResponseWriter, r *http.Request) (*trace.Span, []samplers.JSONMetric, error) {
```

Comment:
Could we make this name a bit more descriptive (and maybe add a docstring)? I had to read the context to figure out what it did.